_posts/2016-10-24-deploy-jekyll-with-travis.md

```diff
@@ -0,0 +1,232 @@
+  skip_cleanup: false
+  on:
+    branch: source
+```
+
+### The deployment script
+
+When the `source` branch is updated and our tests pass, the deployment script is
+executed.
+
+```sh
+#!/bin/bash
+
+# Enable error reporting to the console.
+set -e
+
+# Install bundles if needed.
+bundle check || bundle install
+
+# NPM install if needed.
+. $HOME/.nvm/nvm.sh && nvm install 6.1 && nvm use 6.1
+npm install
+
+# Build the site.
+gulp
+
+# Checkout `master` and remove everything
+git clone https://${GH_TOKEN}@github.com/savaslabs/savaslabs.github.io.git ../savaslabs.github.io.master
+cd ../savaslabs.github.io.master
+git checkout master
+rm -rf *
+
+# Copy generated HTML site from source branch in original repo.
+# Now the `master` branch will contain only the contents of the _site directory.
+cp -R ../savaslabs.github.io/_site/* .
+
+# Make sure we have the updated .travis.yml file so tests won't run on master.
+cp ../savaslabs.github.io/.travis.yml .
+git config user.email ${GH_EMAIL}
+git config user.name "savas-bot"
+
+# Commit and push generated content to `master` branch.
+git status
+git add -A .
+git status
+git commit -a -m "Travis #$TRAVIS_BUILD_NUMBER"
+git push --quiet origin `master` > /dev/null 2>&1
+```
+
+Some important notes:
+
+- Make sure the script file is executable!
+- The last line ensures that our encrypted token won't end up in git logs
+
+## Deployment and debugging
+
+Merging this code into `source` was the easiest way for us to test if our new process worked
+— scary, right? Fortunately, if the build script failed (and it did several
+times until we could get the npm and node versions correct and the ruby gems
+installing properly) nothing got pushed to `master` so nothing was deployed,
+meaning no downtime on our site. We also used
+[Teleconsole](https://www.teleconsole.com/) to debug from inside the Travis
+environment.
+
+To do this, we commented out the `git push` line and added the following to
+`build.sh`:
+
+```sh
+curl https://www.teleconsole.com/get.sh | sh
+teleconsole
+```
+
+This printed a session ID in the Travis CI output. With a couple of commands, we
+could enter the session from our local machines:
+
+```bash
+# Install teleconsole
+$ curl https://www.teleconsole.com/get.sh | sh
+
+# Join session
+$ teleconsole join [session ID]
+```
+
+Being able to look around and run commands within the Travis environment was
+hugely helpful!
+
+## Future work
+
+We'd like to implement some improvements on this process in the future:
+
+- With our current setup, we can't deploy when our tests are failing.
+- Managing ruby and node dependencies has proven difficult across our team and
+within the Travis environment. We've considered Dockerizing our site but that's
+a subject for another post!
+
+## Resources
+
+We'd like to give credit to these helpful posts, which we adapted for our site:
+
+- [From Evgeny Shepelyuk's blog](http://eshepelyuk.github.io/2014/10/28/automate-github-pages-travisci.html)
```

Comment:
Not critical, but I think it would be nice to include the referenced blog post's title.